### PR TITLE
significantly improved support for multi-monitor setups on desktop; other screen, stage, scene improvements

### LINKED ISSUE #509: New windows are sometimes placed offscreen on Windows due to a failure to account for the size of the title bar
There are checks in there that are supposed to prevent this, but they aren't working!

### LINKED ISSUE #603: Triggering a full update on a window with F5 can cause it to relocate
In particular, when using multiple monitors, this change can be quite extreme and put it outside of the bounds of the monitor.

### LINKED ISSUE #947: Crash when getting monitors
### Describe the bug

Probably due to a monitor being plugged in while the app was running.

### How to reproduce

Try plugging a monitor in while an app is running (probably one that it will switch to as the main monitor).

### Example code

_No response_

### Relevant output

```shell
Platform: MacOS
System platform: MacOS
App version: v0.0.9-186-g51e6d9742
Core version: v0.0.9-186-g51e6d9742
Time: 2024-04-04 14:55:17

panic: runtime error: index out of range [2] with length 2

goroutine 1 [running, locked to thread]:
runtime/debug.Stack()
	/opt/homebrew/Cellar/go/1.22.1/libexec/src/runtime/debug/stack.go:24 +0x64
cogentcore.org/core/goosi.HandleRecoverBase({0x106492a40, 0x1401fe7cdb0})
	/Users/kaioreilly/cogent/core/goosi/recover.go:49 +0x38
cogentcore.org/core/gi.HandleRecover({0x106492a40, 0x1401fe7cdb0})
	/Users/kaioreilly/cogent/core/gi/recover.go:32 +0x48
cogentcore.org/core/gi.Wait.func1()
	/Users/kaioreilly/cogent/core/gi/renderwin.go:35 +0x34
panic({0x106492a40?, 0x1401fe7cdb0?})
	/opt/homebrew/Cellar/go/1.22.1/libexec/src/runtime/panic.go:770 +0x124
cogentcore.org/core/goosi/driver/desktop.(*App).GetScreens(0x106f145e0)
	/Users/kaioreilly/cogent/core/goosi/driver/desktop/screen.go:80 +0x1464
cogentcore.org/core/goosi/driver/desktop.(*App).MonitorChange(0x106f145e0, 0x106425420?, 0x1400004db01?)
	/Users/kaioreilly/cogent/core/goosi/driver/desktop/screen.go:39 +0xf8
github.com/go-gl/glfw/v3.3/glfw.goMonitorCB(...)
	/Users/kaioreilly/go/pkg/mod/github.com/go-gl/glfw/v3.3/glfw@v0.0.0-20240307211618-a69d953ea142/monitor.go:52
github.com/go-gl/glfw/v3.3/glfw._Cfunc_glfwWaitEvents()
	_cgo_gotypes.go:2455 +0x30
github.com/go-gl/glfw/v3.3/glfw.WaitEvents()
	/Users/kaioreilly/go/pkg/mod/github.com/go-gl/glfw/v3.3/glfw@v0.0.0-20240307211618-a69d953ea142/window.go:949 +0x1c
cogentcore.org/core/goosi/driver/desktop.(*App).MainLoop(0x106f145e0)
	/Users/kaioreilly/cogent/core/goosi/driver/desktop/app.go:71 +0x10c
cogentcore.org/core/gi.Wait()
	/Users/kaioreilly/cogent/core/gi/renderwin.go:41 +0x58
cogentcore.org/core/gi.(*Stage).Wait(...)
	/Users/kaioreilly/cogent/core/gi/stage.go:311
cogentcore.org/core/gi.(*Body).RunMainWindow(0x1400036c008?)
	/Users/kaioreilly/cogent/core/gi/mainstage.go:41 +0x38
main.main()
	/Users/kaioreilly/cogent/core/examples/demo/demo.go:49 +0x138
```

### Platform

macOS

### LINKED ISSUE #997: be able to specify the monitor which window will be created/displayed [multiple monitors support]
### Describe the feature

 it is quite usual for Desktop application to display different window/content among different monitors when present ,  especially applications developed for professional use , i.e. specify the monitor which will be shown when create a desktop window , if i remembered right , glfw support such a feature , and from code  [MonitorChange](https://github.com/cogentcore/core/blob/b285dfb4e419e83e0e83d5ed30c05b965a64f574/system/driver/desktop/screen.go#L29) and [GetScreens](https://github.com/cogentcore/core/blob/b285dfb4e419e83e0e83d5ed30c05b965a64f574/system/driver/desktop/screen.go#L29) , i think it is quite close to such a feature ? 

thanks 

### Relevant code

_No response_

### LINKED ISSUE #1083: Improve screen-specific settings structure
Repro (on latest Chrome, latest Windows):
* go to https://www.cogentcore.org/
* select `Settings`
* click `+` button next to `Screens`

First, it's completely mysterious what it does.

Second: it adds `Zoom` field with value `10` (10% ?). This makes display almost un-readable.

Changing the `Zoom` value doesn't work. There's no way to remove it. 

I had to delete website data in Chrome to get back to normal state.

### LINKED ISSUE #1329: Allow disabling manual window resizing
### Describe the feature

Since window resizing is broken and makes the application crash (as seen on https://github.com/cogentcore/core/issues/1247) , a quick fix would be disabling window resize.

However, Cogent Core doesn't seem to have any way to prevent the user from resizing the window.

### Relevant code

_No response_

### LINKED ISSUE #1357: Monitor changes are not detected during sleep (MacOS at least)
### Describe the bug

If you connect or disconnect monitors (Screens) while your mac laptop is awake, you reliably get monitor change events per this callback:

```
glfw.SetMonitorCallback(a.MonitorChange)
```

However, if you close your laptop lid, then disconnect an external monitor, then open the lid and sign in, there are no events forthcoming from glfw to let you know that your monitor configuration is not as it was before.

Furthermore, my attempts to google "mac awake from sleep event" did not return anything useful, although admittedly I did not spend an exhaustive amount of time on this.

Basically, we need to somehow get an event when it wakes up (and after you sign in presumably), and then we could check the current list of monitors against whatever was last saved, and then drive a change event if these are different.

Alternatively, we could just poll the monitors every minute or so, using a "last polled" timestamp that is checked in the event loop (including paint events). Typically a wake from sleep scenario would result in a long interval and thus an immediate update upon awaking.

Ok, that is sufficiently actionable that I will go ahead and implement! :)

### How to reproduce

do it.

### Example code

_No response_

### Relevant output

_No response_

### Platform

macOS

### LINKED ISSUE #1359: Window size is not restored properly on 2nd screen at first load of saved geom info, but is thereafter
### Describe the bug

See the reproduction -- all on mac with 2nd monitor.

### How to reproduce

1. move window to laptop monitor from primary external monitor, and set size.
2. close and open laptop, and observe proper memory of size as it flips between laptop when open to external when closed.
3. quit the app, with laptop closed.
4. start app, with laptop closed. then open lid. window goes to the laptop screen, but it is sized much larger. why!?

### Example code

```go
hopefully last issue to fix in #1341..
```

### Relevant output

_No response_

### Platform

macOS

### LINKED ISSUE #1360: Two Screens issues on Windows: not detecting new screen when moving, and crashing on restore from minimize
### Describe the bug

when you move a window to a different screen, it doesn't seem to detect that, and this results in: 1. frame constraint bouncing the window back to the original screen; and 2. not getting the new dpi of the new screen.

also, minimizing works but then it tries to restore and emits a "surface cannot be 0, 0" kind of error -- should be an easy fix.

minimizing and restoring works fine on mac. need to also test on linux.

Both for #1341 

### How to reproduce

do it.

### Example code

_No response_

### Relevant output

_No response_

### Platform

macOS

### PR DESCRIPTION
**Pull request by rcoreilly**
Add screen name to saved window geometry so it prefers to open there, and add option for screen name in opening new windows, in stage and lower level system api. also system.Window.Position() always returns screen-relative position. all working as expected.

Picks up from progress in #1339

* Fixes #603 -- always uses current screen for refresh (UpdateAll).
* Fixes #997 and Fixes #1140 -- can do e.g., `b.NewWindow().SetScreen(1).RunMain()` to specify screen 1 as the default to open a new window.
* Fixes #509 -- added ConstrainToFrame that keeps full window frame in view.
* Fixes #1329 -- added Resizable flag on Stage
* Fixes #1357 -- detects monitor changes after waking from sleep.
* Fixes #1359 -- window Screen() is not always up to date at monitor change events: use wingeom version.
* Fixes #1083 -- screen-specific zoom now used for zoom shortcut and made read-only and fixed comments.
* Fixes #947 -- never saw this in all my testing of plugging and unplugging, but pushed an extra sanity fix anyway.
* Fixes #1360 -- fixed issues in the ConstrainFrame code and detect the -32000 special windows position value.
* May help with #1319 and #1247 based on updates in device config.

### CODE REVIEW

File: core/mainstage.go [hunk 1]
```diff
@@ -46,6 +46,16 @@ func (bd *Body) RunMainWindow() {
 	Wait()
 }
 
+// RunMain runs a MainWindow, starts the app's main loop,
```
Comment (kkoreilly) [inline core/mainstage.go]: Clean up docs.

File: core/mainstage.go [hunk 2]
```diff
@@ -356,18 +366,28 @@ func (st *Stage) newRenderWindow() *renderWindow {
 		Icon:      appIconImages(),
 		Size:      st.Scene.SceneGeom.Size,
 		StdPixels: false,
+		Screen:    st.Screen,
+	}
+	if !st.Resizable {
```
Comment (kkoreilly) [inline core/mainstage.go]: Does this fix #1329? If so, we should mention it and also add support for the same flag in dialog.

File: core/mainstage.go [hunk 3]
```diff
@@ -356,18 +366,28 @@ func (st *Stage) newRenderWindow() *renderWindow {
...
 	}
-	wgp := theWindowGeometrySaver.pref(title, nil)
-	if TheApp.Platform() != system.Offscreen && wgp != nil {
+	wgp, screen := theWindowGeometrySaver.get(title, screenName)
+	if wgp != nil {
 		theWindowGeometrySaver.settingStart()
+		opts.Screen = screen.ScreenNumber
 		opts.Size = wgp.size()
 		opts.Pos = wgp.pos()
+		// fmt.Println("using saved:", screen.Name, wgp.size(), wgp.pos())
```
Comment (kkoreilly) [inline core/mainstage.go]: Remove print statement?

File: core/renderwindow.go [hunk 4]
```diff
@@ -439,6 +439,7 @@ func (w *renderWindow) handleWindowEvents(e events.Event) {
 		w.mains.runDeferred() // note: must be outside of locks in renderWindow
 
 	case events.WindowResize:
+		// fmt.Println("resized")
```
Comment (kkoreilly) [inline core/renderwindow.go]: Remove print.

File: core/renderwindow.go [hunk 5]
```diff
@@ -496,24 +497,25 @@ func (w *renderWindow) handleWindowEvents(e events.Event) {
 			w.gotFocus = false
 			w.sendWinFocusEvent(events.WinFocusLost)
 		case events.ScreenUpdate:
-			if DebugSettings.WinEventTrace {
-				fmt.Printf("Win: %v ScreenUpdate\n", w.name)
-			}
-			w.resized()
+			// if DebugSettings.WinEventTrace {
+			fmt.Println("Win: ScreenUpdate", w.name, screenConfig())
```
Comment (kkoreilly) [inline core/renderwindow.go]: Add back `if Debug...`

File: core/stage.go [hunk 6]
```diff
@@ -97,6 +97,11 @@ type Stage struct { //types:add -setters
 	// [Stage.DisplayTitle] is true.
 	Title string
 
+	// Screen specifies the screen number to open a NewWindow on non-mobile
```
Comment (kkoreilly) [inline core/stage.go]: Clean up docs.

File: core/stage.go [hunk 7]
```diff
@@ -125,6 +130,10 @@ type Stage struct { //types:add -setters
 	// The saved window position and size takes precedence on multi-window platforms.
 	UseMinSize bool
 
+	// Resizable specifies whether a NewWindow on non-mobile platforms can
```
Comment (kkoreilly) [inline core/stage.go]: Clean up docs; add support for dialog resizable as well.

File: core/windowgeometry.go [hunk 8]
```diff
@@ -61,17 +84,23 @@ type windowGeometrySaver struct {
...
+		ws.geometries = make(screenConfigGeometries, 1000)
 		ws.resetCache()
 		ws.filename = "window-geometry"
 		ws.lockSleep = 100 * time.Millisecond
 		ws.saveDelay = 1 * time.Second
 	}
 }
 
+// shouldSave returns whether the window geometry should be saved based on
+// the platform: only for non-mobile (non-web) and not offscreen.
```
Comment (kkoreilly) [inline core/windowgeometry.go]: `only for desktop native`.

File: core/windowgeometry.go [hunk 9]
```diff
@@ -61,17 +84,23 @@ type windowGeometrySaver struct {
 // init does initialization if not yet initialized
 func (ws *windowGeometrySaver) init() {
 	if ws.geometries == nil {
-		ws.geometries = make(windowGeometries, 1000)
+		ws.geometries = make(screenConfigGeometries, 1000)
```
Comment (kkoreilly) [inline core/windowgeometry.go]: Do we really need this much capacity?

File: core/windowgeometry.go [hunk 10]
```diff
@@ -381,15 +428,48 @@ func (ws *windowGeometrySaver) restoreAll() {
 	}
 }
 
-// windowGeometry records the geometry settings used for a given window
+// windowGeometries holds the window geometries for a given window
+// across different screens, and the last screen used.
+type windowGeometries struct {
+	LS string                    // last screen
+	SC map[string]windowGeometry // screen map
```
Comment (kkoreilly) [inline core/windowgeometry.go]: Fix names.

File: core/windowgeometry.go [hunk 11]
```diff
@@ -381,15 +428,48 @@ func (ws *windowGeometrySaver) restoreAll() {
...
-	PX         int
-	PY         int
-	Fullscreen bool
+	DPI float32
+	DPR float32
+	SX  int
+	SY  int
+	PX  int
+	PY  int
+	FS  bool
```
Comment (kkoreilly) [inline core/windowgeometry.go]: Fix names.

File: examples/demo/demo.go [hunk 12]
```diff
@@ -42,7 +42,7 @@ func main() {
 	valueBinding(ts)
 	makeStyles(ts)
 
-	b.RunMainWindow()
+	b.NewWindow().SetScreen(1).RunMain()
```
Comment (kkoreilly) [inline examples/demo/demo.go]: Revert.

File: system/driver/base/app_multi.go [hunk 13]
```diff
@@ -50,7 +50,7 @@ func (a *AppMulti[W]) Screen(n int) *system.Screen {
 	if n < len(a.Screens) {
 		return a.Screens[n]
 	}
-	return nil
+	return a.Screens[0]
```
Comment (kkoreilly) [inline system/driver/base/app_multi.go:53]: A little bit misleading...

Comment (kkoreilly) [reply system/driver/base/app_multi.go:53]: Document it at the very least.

File: system/driver/desktop/window.go [hunk 14]
```diff
@@ -70,18 +70,22 @@ func (w *Window) DeActivate() {
...
+		glfw.WindowHint(glfw.Resizable, glfw.True)
+	}
 	glfw.WindowHint(glfw.Visible, glfw.False) // needed to position
 	glfw.WindowHint(glfw.Focused, glfw.True)
 	// glfw.WindowHint(glfw.ScaleToMonitor, glfw.True)
 	glfw.WindowHint(glfw.ClientAPI, glfw.NoAPI)
 	// glfw.WindowHint(glfw.Samples, 0) // don't do multisampling for main window -- only in sub-render
+	fullscreen := opts.Flags.HasFlag(system.Fullscreen)
 	if fullscreen {
 		glfw.WindowHint(glfw.Maximized, glfw.True)
```
Comment (kkoreilly) [inline system/driver/desktop/window.go:91]: Fullscreen is doing two things at once: maximized and actual fullscreen, which need to be separated; maximized non-fullscreen is a key modality on Windows.

File: system/driver/desktop/window.go [hunk 15]
```diff
@@ -93,13 +97,18 @@ func NewGlfwWindow(opts *system.NewWindowOptions, sc *system.Screen) (*glfw.Wind
...
+		mon = TheApp.Monitors[sc.ScreenNumber]
+	}
+	win, err := glfw.CreateWindow(sz.X, sz.Y, opts.GetTitle(), mon, nil)
 	if err != nil {
 		return win, err
 	}
 
-	if !fullscreen {
-		win.SetPos(opts.Pos.X, opts.Pos.Y)
+	if !opts.Flags.HasFlag(system.Fullscreen) {
```
Comment (kkoreilly) [inline system/driver/desktop/window.go]: Use `!fullscreen`.

File: system/screen.go [hunk 16]
```diff
@@ -188,14 +188,19 @@ func (sc *Screen) ConstrainWinGeom(sz, pos image.Point) (csz, cpos image.Point)
...
-
-	// these are windows-specific special numbers for minimized windows
-	// can be sent here for WinGeom saved geom.
-	if cpos.X == -32000 {
-		cpos.X = 0
-	}
-	if cpos.Y == -32000 {
-		cpos.Y = 50
+	if TheApp.Platform() == Windows {
+		if csz.Y > (scsz.Y - 20) { // leave room for title
```
Comment (kkoreilly) [inline system/screen.go]: There is a programmatic way to determine this with [glfwGetWindowFrameSize](https://www.glfw.org/docs/3.3/window_guide.html#window_size:~:text=The%20above%20functions%20work%20with%20the%20size%20of%20the%20content%20area%2C%20but%20decorated%20windows%20typically%20have%20title%20bars%20and%20window%20frames%20around%20this%20rectangle.%20You%20can%20retrieve%20the%20extents%20of%20these%20with%20glfwGetWindowFrameSize.).

**Review by kkoreilly — CHANGES_REQUESTED** [review]

File: system/driver/base/window_multi.go [hunk 17]
```diff
@@ -39,6 +39,9 @@ type WindowMulti[A system.App, D system.Drawer] struct {
 	// PixSize is the pixel size of the window in raw display dots
 	PixSize image.Point `label:"Pixel size"`
 
+	// FrameSize of the window frame: Min = left, top; Max = right, bottom.
+	FrameSize image.Rectangle
```
Comment (kkoreilly) [inline system/driver/base/window_multi.go]: Use `styles.Sides[int]`.

**Review by kkoreilly — CHANGES_REQUESTED** [review]

File: system/screen.go (hunk 16 above)
Comment (rcoreilly) [reply system/screen.go]: this has been implemented.

File: system/driver/base/window_multi.go (hunk 17 above)
Comment (rcoreilly) [reply system/driver/base/window_multi.go]: done.

**Comment by rcoreilly** [comment]
@kkoreilly this is now done from my perspective!

**Review by kkoreilly — APPROVED** [review]

**Pull request merged**